src/cpp/session/modules/environment/SessionEnvironment.cpp

```diff
@@ -1229,6 +1236,51 @@ void onDetectChanges(module_context::ChangeSource /* source */)
    s_pEnvironmentMonitor->checkForChanges();
 }
 
+namespace {
+
+SEXP inferDebugSrcrefs(boost::shared_ptr<LineDebugState> pLineDebugState)
+{
+   using namespace r::context;
+   
+   // check to see if we have real source references for the currently
+   // executing context
+   SEXP srcref = r::context::globalContext().srcref();
+   if (isValidSrcref(srcref))
+      return srcref;
+   
+   // no source reference available; try to find an appropriate context
+   for (auto it = RCntxt::begin(); it != RCntxt::end(); ++it)
+   {
+      // if we find a CTXT_BROWSER context, try to find its matching CTXT_FUNCTION
+      // these contexts are occasionally created when e.g. stepping into tryCatch
+      if (it->callflag() & CTXT_BROWSER)
+      {
+         SEXP cloenv = it->cloenv();
+         for (; it != RCntxt::end(); it++)
+         {
+            if (it->callflag() & CTXT_FUNCTION)
+            {
+               if (cloenv == it->cloenv())
```

Comment:
Should this continue the search to skip over potential `eval()` frames that have the same cloenv?

IIUC what we do here needs to match what we did when we last updated `pLineDebugState` so that the lastline info is consistent? I'm not sure I understand when and how this happens. Is it in `callFramesAsJson()` called in `onConsolePrompt()` via `enqueContextDepthChangedEvent()`?

Reply:
> Should this continue the search to skip over potential eval() frames that have the same cloenv?

Good catch; I've just added that change. Seems to work well!

> IIUC what we do here needs to match what we did when we last updated pLineDebugState so that the lastline info is consistent? I'm not sure I understand when and how this happens. Is it in callFramesAsJson() called in onConsolePrompt() via enqueContextDepthChangedEvent()?

To be totally frank, I'm not entirely sure 😬  All I know is that "it works" with these changes!